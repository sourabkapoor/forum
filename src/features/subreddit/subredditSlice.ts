import { createSlice } from "@reduxjs/toolkit";
import { DocumentData } from "firebase/firestore";
import { RootState } from "../../redux/store";

interface AuthState {
  communityModalState: boolean;
  communityData: DocumentData;
  userCommunities: DocumentData[];
}

const initialState: AuthState = {
  communityModalState: false,
  communityData: {},
  userCommunities: [],
};

export const selectCommunityData = (state: RootState) =>
  state.subreddit.communityData;

export const selectCommunityModalState = (state: RootState) =>
  state.subreddit.communityModalState;

export const selectUserCommunitiesData = (state: RootState) =>
  state.subreddit.userCommunities;

// export const selectCurrentSubredditData = (state: RootState) =>
//   state.subreddit.currentSubredditData;

const subredditSlice = createSlice({
  name: "subreddit",
  initialState,
  reducers: {
    setCommunityData: (state, action) => {
      state.communityData = action.payload;
    },
    toggleCommunityModalState: (state) => {
      state.communityModalState = !state.communityModalState;
    },
    setUserCommunities: (state, action) => {
      state.userCommunities = action.payload;
    },
  },
});

export const {
  setCommunityData,
  toggleCommunityModalState,
  setUserCommunities,
} = subredditSlice.actions;

export default subredditSlice.reducer;
