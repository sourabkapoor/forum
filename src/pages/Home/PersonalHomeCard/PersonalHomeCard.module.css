.personal-home-card__banner {
    background-image: url("https://www.redditstatic.com/desktop2x/img/id-cards/home-banner@2x.png");
    height: 38px;
    margin: -12px -12px -10px;
}

.personal-home-card__title {
    display: flex;
    align-items: center;
    column-gap: 10px;
}

.personal-home-card__image {
    height: 68px;
    width: 40px;
    background: 40px 68px;
}

.personal-home-card__text {
    font-size: 16px;
    font-weight: 600;
    line-height: 20px;
    color: var(--secondary-color);
}
.personal-home-card__description {
    font-size: 14px;
    font-family: inherit;
    line-height: 21px;
    font-weight: 400;
    word-wrap: break-word;
    max-width: 286px;
}

.personal-home-card__buttons {
    display: flex;
    flex-direction: column;
}

.personal-home-card__button {
    /* background-color: var(--secondary-color); */
    margin-top: 12px;
    /* color: var(--primary-color); */
    font-size: 14px;
    font-weight: 700;
    font-family: inherit;
    letter-spacing: unset;
    text-transform: unset;
    min-height: 32px;
    min-width: 32px;
    padding: 4px 16px;
    border-radius: 9999px;
    box-sizing: border-box;
}
.personal-home-card__button_type_create-post {
    background-color: white;
    color: var(--primary-color);
    border: none;
}
.personal-home-card__button_type_create-community {
    background-color: inherit;
    color: var(--secondary-color);
    border: 1px solid var(--secondary-color);
}

.personal-home-card__button_type_create-post:hover {
    background-color: var(--secondary-color);
    cursor: pointer;
}
.personal-home-card__button_type_create-community:hover {
    background-color: #242424;
    cursor: pointer;
}