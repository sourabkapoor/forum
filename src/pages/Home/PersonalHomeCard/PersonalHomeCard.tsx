import CSSModules from "react-css-modules";
import { useNavigate } from "react-router-dom";
import Card from "../../../components/Card/Card";
import {
  selectCommunityData,
  selectCommunityModalState,
  toggleCommunityModalState,
} from "../../../features/subreddit/subredditSlice";
import { useAppDispatch, useAppSelector } from "../../../hooks/hooks";
import styles from "./PersonalHomeCard.module.css";
import { useEffect, useState } from "react";
import { collection, getDocs, limit, orderBy, query, where } from "firebase/firestore";
import { db, getUserId } from "../../../firebase";
import ReactApexChart from "react-apexcharts";
import moment from "moment";

const PersonalHomeCard: React.FC = () => {
  const [posts, setPosts] = useState<any | []>([])

  const options = {
    chart: {
      height: 350,
      // type: "bar",
    },
    plotOptions: {
      bar: {
        borderRadius: 10,
        dataLabels: {
          position: 'top',
        },
      }
    },
    dataLabels: {
      enabled: true,
      formatter: function (val: any) {
        return val;
      },
      offsetY: -20,
      style: {
        fontSize: '12px',
        colors: ["#304758"]
      }
    },
    xaxis: {
      categories: ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"],
      position: 'bottom',
    },
  }

  const series = [{
    name: 'Posts',
    data: posts.length === 0 ? [0, 0, 0, 0, 0, 0, 0] : posts
  }]

  useEffect(() => {
    async function getLastWeekPosts() {
      const posts = collection(db, "posts");
  
      const q = await query(
        posts,
        where("userId", "==", getUserId())
      );
      await getDocs(q)
      .then(res => {
        const newData = res.docs.map((item) => ({
          ...item.data(), id: item.id
        }))

        const finalPostTimeLine = [0, 0, 0, 0, 0, 0, 0]
        newData.map((post: any) => {
          const day = moment(new Date(post.createdAt.seconds * 1000)).format('ddd')
          const dayCode = moment(new Date(post.createdAt.seconds * 1000)).format('d')
          finalPostTimeLine[parseInt(dayCode)] = finalPostTimeLine[parseInt(dayCode)] + 1
        })
        setPosts(finalPostTimeLine)
      })
    }

    getLastWeekPosts()
  }, [])

  return (
    <Card>
      <h5 style={{marginBottom: 8}}>Your Last Posts</h5>
      {/* Charts */}
      <div>
        {posts.length === 0 && <p style={{textAlign: "center"}}>No data to display</p>}
      </div>

      <ReactApexChart options={options} series={series} type="bar" />

    </Card>
  );
};

export default CSSModules(PersonalHomeCard, styles, {
  allowMultiple: true,
  handleNotFoundStyleName: "ignore",
});
