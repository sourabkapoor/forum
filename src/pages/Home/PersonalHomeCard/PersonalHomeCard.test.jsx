import {
  getByRole,
  queryByPlaceholderText,
  queryByRole,
  queryByText,
  render,
  screen,
} from "@testing-library/react";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";
import each from "jest-each";
import PersonalHomeCard from "./PersonalHomeCard";
import StoreProvider from "../../../redux/provider";
import * as subredditSlice from "../../../features/subreddit/subredditSlice";
import { BrowserRouter } from "react-router-dom";

describe("PersonalHomeCard", () => {
  test("snapshot", () => {
    const { asFragment } = render(
      <BrowserRouter>
        <StoreProvider>
          <PersonalHomeCard />
        </StoreProvider>
      </BrowserRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });
  test("open community modal", async () => {

    const mockDispatch = jest.fn();
    jest.mock("../../../hooks/hooks", () => ({
      ...jest.requireActual("../../../hooks/hooks"),
      useAppDispatch: () => mockDispatch,
    }));

    render(
      <BrowserRouter>
        <StoreProvider>
          <PersonalHomeCard />
        </StoreProvider>
      </BrowserRouter>
    );

    const createCommunityButton = screen.queryByText("Create Community");

    await userEvent.click(createCommunityButton);

    expect(mockDispatch).toBeCalledTimes(1);
  });
  test("navigate to create post page", async () => {

    jest.mock("../../../hooks/hooks", () => ({
      ...jest.requireActual("../../../hooks/hooks"),
      useAppSelector: () => jest.fn().mockReturnValue("name"),
    }));

    const mockNavigate = jest.fn();
    jest.mock("react-router-dom", () => ({
      ...jest.requireActual("react-router-dom"),
      useNavigate: () => mockNavigate,
    }));

    render(
      <BrowserRouter>
        <StoreProvider>
          <PersonalHomeCard />
        </StoreProvider>
      </BrowserRouter>
    );

    const createPostButton = screen.queryByText("Create Post")

    await userEvent.click(createPostButton)

    expect(mockNavigate).toHaveBeenCalledTimes(1)
  })

});
