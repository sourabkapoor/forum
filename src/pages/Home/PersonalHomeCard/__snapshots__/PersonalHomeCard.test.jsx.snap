// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonalHomeCard snapshot 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <div
      class="personal-home-card"
    >
      <div
        class="personal-home-card__banner"
      />
      <div
        class="personal-home-card__title"
      >
        <img
          alt="redditor"
          class="personal-home-card__image"
          src="https://www.redditstatic.com/desktop2x/img/id-cards/snoo-home@2x.png"
        />
        <span
          class="personal-home-card__text"
        >
          Home
        </span>
      </div>
      <p
        class="personal-home-card__description"
      >
        Your personal Reddit frontpage. Come here to check in with your favorite communities.
      </p>
      <div
        class="personal-home-card__buttons"
      >
        <button
          class="personal-home-card__button personal-home-card__button_type_create-post"
        >
          Create Post
        </button>
        <button
          class="personal-home-card__button personal-home-card__button_type_create-community"
        >
          Create Community
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
